### PR TITLE
[Merged by Bors] - feat: Positivity extension for `NNRat.num`/`NNRat.den`

### PR DESCRIPTION
**Pull request by YaelDillies**
From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 1]
```diff
@@ -420,6 +421,48 @@ def evalAscFactorial : PositivityExt where eval {u α} _ _ e := do
...
+open NNRat
+
+private alias ⟨_, NNRat.num_pos_of_pos⟩ := num_pos
+private alias ⟨_, NNRat.num_ne_zero_of_ne_zero⟩ := num_ne_zero
+
+/-- The `positivity` extension which identifies expressions of the form `NNRat.num q`,
+such that `positivity` successfully recognises `q`. -/
+@[positivity NNRat.num _]
+def evalNNRatNum : PositivityExt where eval {u α} _ _ e := do
+  if let 0 := u then -- lean4#3060 means we can't combine this with the match below
```
Comment (eric-wieser) [inline Mathlib/Tactic/Positivity/Basic.lean:434]: This issue is closed, there is no need for this workaround.

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 2]
```diff
@@ -431,29 +431,25 @@ private alias ⟨_, NNRat.num_ne_zero_of_ne_zero⟩ := num_ne_zero
...
-    | ~q(ℕ), ~q(NNRat.den $a) =>
-      assumeInstancesCommute
-      return .positive q(den_pos $a)
-    | _, _ => throwError "not NNRat.num"
-  else throwError "not NNRat.num"
+  match u, α, e with
+  | 0, ~q(ℕ), ~q(NNRat.den $a) =>
+    assumeInstancesCommute
+    return .positive q(den_pos $a)
+  | _, _, _ => throwError "not NNRat.num"
```
Comment (eric-wieser) [inline Mathlib/Tactic/Positivity/Basic.lean:452]:
```suggestion
  | _, _, _ => throwError "not NNRat.den"
```

File: Mathlib/Tactic/Positivity/Basic.lean [hunk 3]
```diff
@@ -420,6 +421,44 @@ def evalAscFactorial : PositivityExt where eval {u α} _ _ e := do
     pure (.positive q(Nat.ascFactorial_pos $n $k))
   | _, _, _ => throwError "failed to match Nat.ascFactorial"
 
+section NNRat
+open NNRat
+
+private alias ⟨_, NNRat.num_pos_of_pos⟩ := num_pos
+private alias ⟨_, NNRat.num_ne_zero_of_ne_zero⟩ := num_ne_zero
```
Comment (eric-wieser) [inline Mathlib/Tactic/Positivity/Basic.lean:428]: Do you think these should just be public (after checking they end up in the right namespace)?

bors d+

Comment (YaelDillies) [reply Mathlib/Tactic/Positivity/Basic.lean:428]: My stance on whether one sided implications of unconditional iff lemmas without potential for dot notation should exist is "not really".

Btw the reason I added the `NNRat` namespace, as opposed to the
```
open Rat

private alias ⟨_, num_pos_of_pos⟩ := num_pos
private alias ⟨_, num_ne_zero_of_ne_zero⟩ := num_ne_zero
```
just below, is that both pairs of aliases name-conflict! The lemmas aren't autogenerated to guarantee uniqueness.

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**